### PR TITLE
Cut over settings registration to a pull model

### PR DESCRIPTION
**Pull request by s1monw**
Today we have a push model for registering basically anything. All our extension points
are defined on modules which we pass in to plugins. This is harder to maintain and adds
unnecessary dependencies on the modules itself. This change moves towards a pull model
where the plugin offers a getter kind of method to get the extensions. This will also
help in the future if we need to pass dependencies to the extension points which can
easily be defined on the method as arguments if a pull model is used.

### CODE REVIEW

**Comment by s1monw** [comment]
@rjernst can you take a look?

File: core/src/main/java/org/elasticsearch/common/settings/Setting.java [hunk 1]
```diff
@@ -248,6 +253,13 @@ public boolean isDeprecated() {
     }
 
     /**
+     * Returns <code>true</code> if this setting is shared with more than one other module or pluging, otherwise <code>false</code>
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/Setting.java]: pluging -> plugin

File: core/src/main/java/org/elasticsearch/common/settings/SettingsModule.java [hunk 2]
```diff
@@ -200,4 +203,16 @@ private void validateTribeSettings(Settings settings, ClusterSettings clusterSet
...
+
+    public Settings getSettings() {
+        return settings;
+    }
+
+    public IndexScopedSettings getIndexScopedSettings() {
+        return indexScopedSettings;
+    }
+
+    public ClusterSettings getClusterSettings() {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/SettingsModule.java]: Nit, and nothing to do with this PR but just something I noticed: we should make this terminology clearer, for example either using "scoped" in all of these or none of them? And perhaps having "node" in the first one, like getNodeSettings()? Just an idea for future work.

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 3]
```diff
@@ -250,13 +254,29 @@ protected Node(Environment tmpEnv, Version version, Collection<Class<? extends P
...
-            threadPoolModule.prepareSettings(settingsModule);
-
+            // TODO - we have to move up the script registration to run first and then
+            // we can validate and build all settings classes before we pass the settings anywhere.
+            // this essentially means that we can start constructing the node level service without
+            // guice and define our own extension points. Without this we won't be able to get rid of
+            // guice.
+            // we alread have our own construction of BigArrays, ThreadPool which are one of the
+            // most used resources. Yet, there is also CircuitBreakerService which is needed
+            // in many places and therefor we have to move it up since it registers settings
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/node/Node.java]: therefor -> therefore

**Comment by rjernst** [comment]
This LGTM. It is an awesome example in starting to remove guice from plugins.

One thought is about the new "shared" setting property. I wonder if it would be simpler to enforce that a single plugin, or ES core, does not register the same setting twice, and otherwise just enforce, similar to mappings, that the setting matches an already registered setting (so two plugins cannot try to register the same setting name with different properties).

**Comment by s1monw** [comment]
> One thought is about the new "shared" setting property. I wonder if it would be simpler to enforce that a single plugin, or ES core, does not register the same setting twice, and otherwise just enforce, similar to mappings, that the setting matches an already registered setting (so two plugins cannot try to register the same setting name with different properties).

I think it's much simpler to explicitly mark the ones that can be duplicates since it has to be done on both ends... (I guess that means my code is wrong :) I should look up the already registered setting and check if its also shared. I think this would protect us from others overriding any settings on us?

**Comment by rjernst** [comment]
My concern is a setting that we don't mark as shared, but then other external plugin wants to read that setting. Now they can't do it until we update our code?

**Comment by s1monw** [comment]
> My concern is a setting that we don't mark as shared, but then an external plugin wants to read that setting. Now they can't do it until we update our code?


they can't register it twice that is true. It think we shouldn't design for this exception it's really only there for the S3 / AWS separation, should we really overdesign this?

File: core/src/main/java/org/elasticsearch/common/settings/SettingsModule.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/settings/SettingsModule.java]: yeah I had this idea a couple of times :)

**Pull request merged**